packages/core-data/src/reducer.js

```diff
@@ -489,30 +493,31 @@ export function undo( state = UNDO_INITIAL_STATE, action ) {
 		};
 	};
 
-	const appendEditsToStack = (
+	const appendEditToStack = (
 		stack = [],
-		{ kind, name, recordId, edits }
+		{ kind, name, recordId, property, from, to }
 	) => {
 		const existingEditIndex = stack?.findIndex(
-			( { kind: k, name: n, recordId: r } ) => {
-				return k === kind && n === name && r === recordId;
+			( { kind: k, name: n, recordId: r, property: p } ) => {
+				return (
+					k === kind && n === name && r === recordId && p === property
+				);
 			}
 		);
-
-		const nextStack = stack.filter(
-			( _, index ) => index !== existingEditIndex
-		);
-		nextStack.push( {
-			kind,
-			name,
-			recordId,
-			edits: {
-				...( existingEditIndex !== -1
-					? stack[ existingEditIndex ].edits
-					: {} ),
-				...edits,
-			},
-		} );
+		const nextStack = [ ...stack ];
+		if ( existingEditIndex !== -1 ) {
+			// If the edit is already in the stack leave the initial "from" value.
+			nextStack[ existingEditIndex ].to = to;
```

Comment:
This is modifying also the _previous_ stack, which should remain immutable. The `[...stack]` statement does only a shadow one-level copy, but we're doing a modification at the second level, where both stack share the data.

Reply:
There is a change in how `appendEditsToStack` merges new edits into existing ones. The old version pushed the merged edit at the end, and removed the existing one. The new version modifies the existing one, leaving it at the original position. Is that intentional or is it a mistake?

Reply:
> The new version modifies the existing one, leaving it at the original position. Is that intentional or is it a mistake?

It doesn't matter, I just went for the simplest here. I'll fix the mutation though. Good catch.